### PR TITLE
Optimise alignment/part4 add traceback matrix

### PR DESCRIPTION
**Pull request by rrahn**
part of #1908 

Adds a trace matrix in the new and optimised alignment design.

### CODE REVIEW

File: include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+{
+private:
+    //!\brief A lightweight representation of a single column from the complete matrix.
+    using single_trace_column_type = std::span<trace_t>;
+    //!\brief The type of the zipped score column.
+    using matrix_column_type = decltype(views::zip(std::declval<single_trace_column_type>(),
+                                                   std::declval<physical_column_t &>(),
+                                                   std::declval<virtual_column_t &>()));
+    //!\brief The column type as value type.
+    using matrix_column_value_t = std::vector<std::ranges::range_value_t<matrix_column_type>>;
```
Comment (marehr) [inline include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp:169]: Just hypothetical speaking, do we need this? Does a user know how to use the value type? What are the guarantees of the value type? Does it really require a deep copy?

Maybe we should also turn-around the semantics, that the view is the reference (because it behaves much closer to a &) and that we need a special value type.

We have some places now that use views as reference type, and we might want to generalise it.

I have something like this in mind:

```c++
template <std::views::view view_t>
struct basic_view_iterator_value_type : std::vector<std::ranges::range_value_t<view_t>>
{
    using reference = view_t;
    using value_type = std::vector<std::ranges::range_value_t<view_t>>;

    basic_view_iterator_value_type(view_t view)
    { std::ranges::copy(column, std::cpp20::back_inserter(target)); }
}
```

(We force with this, that the common type is the value_type, not sure if we want this)

or

```c++
template <std::views::view view_t>
struct basic_view_iterator_reference_type : view_t
{

    using reference = view_t;
    using value_type = std::vector<std::ranges::range_value_t<view_t>>;

    constexpr operator value_type() const /*memo: const/non-const may need to be considered separately */
    {
        value_type target{};
        std::ranges::copy(*this, std::cpp20::back_inserter(target));
        return target;
    }
}
```

(We force with this, that the common type is the reference_type, not sure if we want this)

File: include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+     * \{
+     */
+    //!\brief Returns the iterator pointing to the first column.
+    iterator begin()
+    {
+        return iterator{*this, 0u};
+    }
+
+    //!\brief This score matrix is not const-iterable.
+    iterator begin() const = delete;
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp:133]: I think, it tells you that the requested function is deleted rather than telling you that it can't call the function and listing all tried candidates.

File: include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    {
+        ++current_column_id;
+        return *this;
+    }
+
+    //!\brief Move `this` to the next column.
+    void operator++(int)
+    {
+        ++(*this);
+    }
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp:245]: yes, the legacy input iterator must return a copy of the value before the increment.

File: include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+     *
+     * Basic exception guarantee. Might throw std::bad_alloc on resizing the internal matrices.
+     */
+    template <std::integral column_index_t, std::integral row_index_t>
+    void resize(column_index_type<column_index_t> const column_count,
+                row_index_type<row_index_t> const row_count)
+    {
+        this->column_count = column_count.get();
+        this->row_count = row_count.get();
+        complete_matrix.resize(number_rows{this->row_count}, number_cols{this->column_count});
```
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp:118]: Like in the other PR. When all these changes are applied, then I would make a new PR to make the resize interface alike.

File: include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rrahn) [reply include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp:133]: This error:
```
<source>:26:28: error: use of deleted function 'int foo::begin() const'

   26 |     std::as_const(f).begin();

      |                            ^

<source>:10:9: note: declared here

   10 |     int begin() const = delete;

      | 
```

vs this error:

```
<source>:27:28: error: passing 'std::add_const_t<bar>' {aka 'const bar'} as 'this' argument discards qualifiers [-fpermissive]

   27 |     std::as_const(b).begin();

      |                            ^

<source>:15:9: note:   in call to 'int bar::begin()'

   15 |     int begin()

      |         ^~~~~

Compiler returned: 1
```

From https://godbolt.org/z/a0oG8F

File: include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp (hunk 1 above)
Comment (rrahn) [reply include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp:169]: I think in both cases the common type is the value type and also correct in this case. I mean no matter how you put it, you always convert the view into a vector.
The issue with the to_simd_view and the proposed solution with the reference proxy was, what was also hammering in my head. I solved it now like this, but I honestly prefer the special value_type with a constructor from the view more.
Simply, because in most cases I don't use the value type but the reference type and would like to avoid the indirection with the proxy.

> Just hypothetical speaking, do we need this? Does a user know how to use the value type? What are the guarantees of the value type? Does it really require a deep copy?

Basically no! The class itself is no user class and only used inside of the algorithm. So we could basically ignore that value type and reference type are not the same. However, if someone does use it and expect a correct implementation then:
```cpp
iter_value_type v = *it;
```
is weird, since it basically means self-assignment. Because the returned view itself is, as you mentioned it, also just like `&` to some memory.

File: include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+ * \tparam trace_t The type of the trace; must be the same as seqan3::detail::trace_directions.
+ *
+ * \details
+ *
+ * In the default trace back implementation we allocate the entire matrix using one byte per cell to store the
+ * seqan3::detail::trace_directions.
+ *
+ * ### Range interface
+ *
+ * The matrix offers a input range interface over the columns of the matrix. Dereferencing the iterator will return
```
Comment (smehringer) [inline include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp:44]:
```suggestion
 * The matrix offers an input range interface over the columns of the matrix. Dereferencing the iterator will return
```
